image-prune

```diff
@@ -24,125 +24,88 @@ import argparse
-             for entry in entries]
-    return [link for link in links if link.endswith(".qcow2") or link.endswith(".iso")]
-
-
-def get_image_names(quiet=False, open_pull_requests=True, offline=False):
-    """Return all image names used by all branches and optionally in open pull requests
-    """
-    images = set()
-    # iterate over visible refs (mostly branches)
-    # this hinges on being in the top level directory of the the git checkout
-    (refs, considerable) = get_refs(open_pull_requests, offline)
-    # list images present in each branch / pull request
-    for name, ref in refs.items():
-        if not quiet:
-            sys.stderr.write("Considering images from {0} ({1})\n".format(name, ref))
-        for link in get_image_links(ref, "images"):
-            if ref in considerable:
-                for consider in considerable[ref]:
-                    if link.startswith(consider):
-                        images.add(link)
-            else:
-                images.add(link)
-
-    return images
-
-
-def get_keepers(quiet=False, open_pull_requests=True, offline=False, checkout_only=False):
-    if checkout_only:
-        targets = set()
-    else:
-        targets = get_image_names(quiet, open_pull_requests, offline)
+def git(*args, cwd='.'):
+    return subprocess.check_output(['git', *args], text=True, cwd=cwd)
+
+
+def get_images_in_commit(commit, cwd='.'):
+    """Returns all images pointed to by a symlink in images/ from the given commit"""
+
+    logger.debug('Keeping all images in %s...', commit)
+    for line in git('cat-file', '-p', f'{commit}:images/', cwd=cwd).splitlines():
+        mode, kind, blob, _name = line.split()
+        if mode == '120000' and kind == 'blob':
+            image = git('cat-file', 'blob', blob, cwd=cwd)
+            logger.debug('  - %s', image)
+            yield image
+
+
+def get_images_in_branches(merge_target, *branch_patterns, cwd='.'):
+    """Returns all images introduced by branches, relative to merge_target"""
+
+    # KISS: just get a diff and grab anything that looks like an image update
+    for ref_line in git('for-each-ref', *branch_patterns, cwd=cwd).splitlines():
+        _rev, _kind, ref = ref_line.split()
+        logger.debug('Considering images changed in %s...', ref)
+        for patch_line in git('diff', f'{merge_target}...{ref}', '--', 'images/', cwd=cwd).splitlines():
+            if match := re.search(r'^\+([-0-9a-z]+-[0-9a-f]+.(qcow2|iso))', patch_line):
+                image = match.group(1)
+                logger.debug('  - %s', image)
+                yield image
+
+
+def get_remote_images():
+    """Returns all images currently used by PRs and origin branches on GitHub"""
+
+    api = github.GitHub()
+    logger.debug('Querying open PRs on %s', api.repo)
+    open_prs = [pr['number'] for pr in api.pulls()]
+    logger.debug('Open PRs: %s', open_prs)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        git('init', cwd=tmpdir)
```

Comment:
This could use a comment about the "why?". I suspect you do this to avoid cluttering the local checkout with lots of extra fetches, and avoid piling up cruft in .git?

Reply:
I added a comment.